File: android/DKParkingSDK/src/main/kotlin/com/dkparking/sdk/backend/TelemetryUploader.kt
// TelemetryUploader.kt
// DK Parking Engine SDK — Android Version 1
// Async telemetry batch uploader — SS-10 per SYSTEM_ARCHITECTURE.md
// Per privacy_and_telemetry_spec.md: no images, no GPS, no user identifiers.
// Upload MUST NOT block or delay the legal evaluation path.

package com.dkparking.sdk.backend

import com.dkparking.sdk.core.ParkingEvaluationResult
import com.dkparking.sdk.core.VersionRefs
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL
import java.time.Instant
import java.util.concurrent.CopyOnWriteArrayList

// MARK: - Telemetry event model

@Serializable
data class TelemetryEvent(
    @SerialName("event_type")              val eventType: String,
    @SerialName("evaluation_id")           val evaluationId: String? = null,
    @SerialName("session_id")              val sessionId: String? = null,
    @SerialName("timestamp_utc")           val timestampUtc: String,
    @SerialName("decision_state")          val decisionState: String? = null,
    @SerialName("refusal_reasons")         val refusalReasons: List<String>? = null,
    @SerialName("rule_family")             val ruleFamily: String? = null,
    @SerialName("confidence_score")        val confidenceScore: Double? = null,
    @SerialName("measured_distance_m")     val measuredDistanceM: Double? = null,
    @SerialName("signed_margin_m")         val signedMarginM: Double? = null,
    @SerialName("total_estimated_error_m") val totalEstimatedErrorM: Double? = null,
    @SerialName("boundary_provenance")     val boundaryProvenance: String? = null,
    @SerialName("ar_metric_scale_score")   val arMetricScaleScore: Double? = null,
    @SerialName("ar_plane_stability_score") val arPlaneStabilityScore: Double? = null,
    @SerialName("in_near_threshold_zone")  val inNearThresholdZone: Boolean? = null,
    @SerialName("sdk_version")             val sdkVersion: String? = null,
    @SerialName("policy_version")          val policyVersion: String? = null,
    @SerialName("dataset_version")         val datasetVersion: String? = null,
    @SerialName("dataset_region_id")       val datasetRegionId: String? = null,
    @SerialName("model_version")           val modelVersion: String? = null,
    @SerialName("platform")               val platform: String = "android",
    @SerialName("os_version")             val osVersion: String? = null,
    @SerialName("evaluation_count")       val evaluationCount: Int? = null,
    @SerialName("refusal_count")          val refusalCount: Int? = null
)

@Serializable
private data class BatchPayload(
    val platform: String,
    val events: List<TelemetryEvent>
)

// MARK: - Uploader

/**
 * Thread-safe async telemetry uploader.
 * Enqueue events with enqueue(). Call flush() on session end.
 * Upload failures are silently ignored — must not affect the legal evaluation path.
 */
class TelemetryUploader(
    private val backendBaseUrl: String,
    private val apiKey: String
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val queue = CopyOnWriteArrayList<TelemetryEvent>()
    private val maxBatchSize = 20
    private val json = Json { ignoreUnknownKeys = true }

    /** Enqueue an event. Non-blocking. */
    fun enqueue(event: TelemetryEvent) {
        queue.add(event)
        if (queue.size >= maxBatchSize) {
            flush()
        }
    }

    /** Flush all queued events. Call on session end. */
    fun flush() {
        if (queue.isEmpty()) return
        val batch = ArrayList(queue.take(maxBatchSize))
        queue.removeAll(batch.toSet())
        scope.launch { uploadBatch(batch) }
    }

    private fun uploadBatch(events: List<TelemetryEvent>) {
        try {
            val payload = BatchPayload(platform = "android", events = events)
            val body = json.encodeToString(payload).toByteArray(Charsets.UTF_8)
            val url = URL("$backendBaseUrl/api/v1/telemetry/batch")
            val conn = url.openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/json")
            conn.setRequestProperty("X-API-Key", apiKey)
            conn.doOutput = true
            conn.connectTimeout = 10_000
            conn.readTimeout = 15_000
            conn.outputStream.use { it.write(body) }
            val code = conn.responseCode
            if (code != 200) {
                println("[TelemetryUploader] server rejected batch: $code")
            }
            conn.disconnect()
        } catch (e: Exception) {
            // Telemetry failure must NOT propagate
            println("[TelemetryUploader] upload failed (ignored): ${e.message}")
        }
    }

    private fun <E> List<E>.take(n: Int): List<E> = subList(0, minOf(n, size))
}

// MARK: - Convenience factory extensions

fun TelemetryEvent.Companion.evaluationCompleted(
    result: ParkingEvaluationResult,
    osVersion: String
) = TelemetryEvent(
    eventType             = "evaluation_completed",
    evaluationId          = result.evaluationId,
    timestampUtc          = Instant.now().toString(),
    decisionState         = result.decisionState.rawValue,
    refusalReasons        = result.refusalReasons.map { it.rawValue },
    ruleFamily            = result.ruleFamily.rawValue,
    confidenceScore       = result.measurement?.confidenceScore,
    measuredDistanceM     = result.measurement?.measuredDistanceM,
    signedMarginM         = result.measurement?.signedMarginM,
    totalEstimatedErrorM  = result.measurement?.totalEstimatedErrorM,
    boundaryProvenance    = result.measurement?.boundaryProvenance?.rawValue,
    arMetricScaleScore    = result.captureQuality.arMetricScaleScore,
    arPlaneStabilityScore = result.captureQuality.arPlaneStabilityScore,
    inNearThresholdZone   = result.measurement?.inNearThresholdZone,
    sdkVersion            = result.versionRefs.sdkVersion,
    policyVersion         = result.versionRefs.policyVersion,
    datasetVersion        = result.versionRefs.datasetVersion,
    datasetRegionId       = result.versionRefs.datasetRegionId,
    modelVersion          = result.versionRefs.modelVersion,
    platform              = "android",
    osVersion             = osVersion
)

fun TelemetryEvent.Companion.sessionStarted(
    sessionId: String,
    versionRefs: VersionRefs,
    osVersion: String
) = TelemetryEvent(
    eventType       = "session_started",
    sessionId       = sessionId,
    timestampUtc    = Instant.now().toString(),
    sdkVersion      = versionRefs.sdkVersion,
    policyVersion   = versionRefs.policyVersion,
    datasetVersion  = versionRefs.datasetVersion,
    datasetRegionId = versionRefs.datasetRegionId,
    modelVersion    = versionRefs.modelVersion,
    platform        = "android",
    osVersion       = osVersion
)

fun TelemetryEvent.Companion.sessionEnded(
    sessionId: String,
    evaluationCount: Int,
    refusalCount: Int,
    versionRefs: VersionRefs
) = TelemetryEvent(
    eventType       = "session_ended",
    sessionId       = sessionId,
    timestampUtc    = Instant.now().toString(),
    sdkVersion      = versionRefs.sdkVersion,
    policyVersion   = versionRefs.policyVersion,
    datasetVersion  = versionRefs.datasetVersion,
    datasetRegionId = versionRefs.datasetRegionId,
    modelVersion    = versionRefs.modelVersion,
    platform        = "android",
    evaluationCount = evaluationCount,
    refusalCount    = refusalCount
)

File: ios/DKParkingSDK/Sources/DKParkingSDK/Backend/TelemetryUploader.swift
// TelemetryUploader.swift
// DK Parking Engine SDK — iOS Version 1
// Async telemetry batch uploader — SS-10 per SYSTEM_ARCHITECTURE.md
// Per privacy_and_telemetry_spec.md: no images, no GPS, no user identifiers.
// Upload MUST NOT block or delay the legal evaluation path.

import Foundation

// MARK: - Telemetry event models

public struct TelemetryEvent: Codable {
    public let eventType: String
    public let evaluationId: String?
    public let sessionId: String?
    public let timestampUtc: String
    public let decisionState: String?
    public let refusalReasons: [String]?
    public let ruleFamily: String?
    public let confidenceScore: Double?
    public let measuredDistanceM: Double?
    public let signedMarginM: Double?
    public let totalEstimatedErrorM: Double?
    public let boundaryProvenance: String?
    public let arMetricScaleScore: Double?
    public let arPlaneStabilityScore: Double?
    public let inNearThresholdZone: Bool?
    public let sdkVersion: String?
    public let policyVersion: String?
    public let datasetVersion: String?
    public let datasetRegionId: String?
    public let modelVersion: String?
    public let platform: String
    public let osVersion: String?
    public let evaluationCount: Int?
    public let refusalCount: Int?

    enum CodingKeys: String, CodingKey {
        case eventType             = "event_type"
        case evaluationId          = "evaluation_id"
        case sessionId             = "session_id"
        case timestampUtc          = "timestamp_utc"
        case decisionState         = "decision_state"
        case refusalReasons        = "refusal_reasons"
        case ruleFamily            = "rule_family"
        case confidenceScore       = "confidence_score"
        case measuredDistanceM     = "measured_distance_m"
        case signedMarginM         = "signed_margin_m"
        case totalEstimatedErrorM  = "total_estimated_error_m"
        case boundaryProvenance    = "boundary_provenance"
        case arMetricScaleScore    = "ar_metric_scale_score"
        case arPlaneStabilityScore = "ar_plane_stability_score"
        case inNearThresholdZone   = "in_near_threshold_zone"
        case sdkVersion            = "sdk_version"
        case policyVersion         = "policy_version"
        case datasetVersion        = "dataset_version"
        case datasetRegionId       = "dataset_region_id"
        case modelVersion          = "model_version"
        case platform
        case osVersion             = "os_version"
        case evaluationCount       = "evaluation_count"
        case refusalCount          = "refusal_count"
    }
}

// MARK: - Uploader

public final class TelemetryUploader {

    private let backendBaseURL: URL
    private let apiKey: String
    private var queue: [TelemetryEvent] = []
    private let maxBatchSize = 20
    private let uploadQueue = DispatchQueue(label: "com.dkparking.telemetry", qos: .background)
    private var uploadTask: URLSessionDataTask?

    public init(backendBaseURL: URL, apiKey: String) {
        self.backendBaseURL = backendBaseURL
        self.apiKey = apiKey
    }

    // MARK: - Enqueue

    /// Enqueue an event. Non-blocking. Thread-safe.
    public func enqueue(_ event: TelemetryEvent) {
        uploadQueue.async { [weak self] in
            guard let self else { return }
            self.queue.append(event)
            if self.queue.count >= self.maxBatchSize {
                self.flushInternal()
            }
        }
    }

    /// Flush all queued events immediately. Call on session end.
    public func flush() {
        uploadQueue.async { [weak self] in
            self?.flushInternal()
        }
    }

    // MARK: - Flush (internal, runs on uploadQueue)

    private func flushInternal() {
        guard !queue.isEmpty else { return }
        let batch = Array(queue.prefix(maxBatchSize))
        queue.removeFirst(min(batch.count, queue.count))

        let payload = BatchPayload(platform: "ios", events: batch)
        guard let data = try? JSONEncoder().encode(payload) else { return }

        var request = URLRequest(url: backendBaseURL.appendingPathComponent("api/v1/telemetry/batch"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.httpBody = data
        request.timeoutInterval = 15

        URLSession.shared.dataTask(with: request) { [weak self] _, response, error in
            if let error = error {
                // Telemetry failure must NOT propagate — log and discard
                print("[TelemetryUploader] upload failed (ignored): \(error.localizedDescription)")
                return
            }
            if let http = response as? HTTPURLResponse, http.statusCode != 200 {
                print("[TelemetryUploader] server rejected batch: \(http.statusCode)")
            }
        }.resume()
    }

    private struct BatchPayload: Codable {
        let platform: String
        let events: [TelemetryEvent]
    }
}

// MARK: - Convenience factory methods

extension TelemetryEvent {

    static func evaluationCompleted(
        result: ParkingEvaluationResult,
        osVersion: String
    ) -> TelemetryEvent {
        TelemetryEvent(
            eventType:             "evaluation_completed",
            evaluationId:          result.evaluationId,
            sessionId:             nil,
            timestampUtc:          ISO8601DateFormatter().string(from: Date()),
            decisionState:         result.decisionState.rawValue,
            refusalReasons:        result.refusalReasons.map { $0.rawValue },
            ruleFamily:            result.ruleFamily.rawValue,
            confidenceScore:       result.measurement?.confidenceScore,
            measuredDistanceM:     result.measurement?.measuredDistanceM,
            signedMarginM:         result.measurement?.signedMarginM,
            totalEstimatedErrorM:  result.measurement?.totalEstimatedErrorM,
            boundaryProvenance:    result.measurement?.boundaryProvenance.rawValue,
            arMetricScaleScore:    result.captureQuality.arMetricScaleScore,
            arPlaneStabilityScore: result.captureQuality.arPlaneStabilityScore,
            inNearThresholdZone:   result.measurement?.inNearThresholdZone,
            sdkVersion:            result.versionRefs.sdkVersion,
            policyVersion:         result.versionRefs.policyVersion,
            datasetVersion:        result.versionRefs.datasetVersion,
            datasetRegionId:       result.versionRefs.datasetRegionId,
            modelVersion:          result.versionRefs.modelVersion,
            platform:              "ios",
            osVersion:             osVersion,
            evaluationCount:       nil,
            refusalCount:          nil
        )
    }

    static func sessionStarted(sessionId: String, versionRefs: VersionRefs, osVersion: String) -> TelemetryEvent {
        TelemetryEvent(
            eventType: "session_started", evaluationId: nil, sessionId: sessionId,
            timestampUtc: ISO8601DateFormatter().string(from: Date()),
            decisionState: nil, refusalReasons: nil, ruleFamily: nil,
            confidenceScore: nil, measuredDistanceM: nil, signedMarginM: nil,
            totalEstimatedErrorM: nil, boundaryProvenance: nil,
            arMetricScaleScore: nil, arPlaneStabilityScore: nil, inNearThresholdZone: nil,
            sdkVersion: versionRefs.sdkVersion, policyVersion: versionRefs.policyVersion,
            datasetVersion: versionRefs.datasetVersion, datasetRegionId: versionRefs.datasetRegionId,
            modelVersion: versionRefs.modelVersion, platform: "ios", osVersion: osVersion,
            evaluationCount: nil, refusalCount: nil
        )
    }

    static func sessionEnded(sessionId: String, evaluationCount: Int, refusalCount: Int, versionRefs: VersionRefs) -> TelemetryEvent {
        TelemetryEvent(
            eventType: "session_ended", evaluationId: nil, sessionId: sessionId,
            timestampUtc: ISO8601DateFormatter().string(from: Date()),
            decisionState: nil, refusalReasons: nil, ruleFamily: nil,
            confidenceScore: nil, measuredDistanceM: nil, signedMarginM: nil,
            totalEstimatedErrorM: nil, boundaryProvenance: nil,
            arMetricScaleScore: nil, arPlaneStabilityScore: nil, inNearThresholdZone: nil,
            sdkVersion: versionRefs.sdkVersion, policyVersion: versionRefs.policyVersion,
            datasetVersion: versionRefs.datasetVersion, datasetRegionId: versionRefs.datasetRegionId,
            modelVersion: versionRefs.modelVersion, platform: "ios", osVersion: nil,
            evaluationCount: evaluationCount, refusalCount: refusalCount
        )
    }
}
